src/libsystemd-network/sd-radv.c

```diff
@@ -987,3 +1062,84 @@ int sd_radv_route_prefix_set_lifetime(sd_radv_route_prefix *p, uint64_t lifetime
 
         return 0;
 }
+
+int sd_radv_pref64_prefix_new(sd_radv_pref64_prefix **ret) {
+        sd_radv_pref64_prefix *p;
+
+        assert_return(ret, -EINVAL);
+
+        p = new(sd_radv_pref64_prefix, 1);
+        if (!p)
+                return -ENOMEM;
+
+        *p = (sd_radv_pref64_prefix) {
+                .n_ref = 1,
+
+                .opt.type = RADV_OPT_PREF64,
+                .opt.length = 2,
+                .prefixlen = 96,
+                .lifetime_usec = RADV_DEFAULT_VALID_LIFETIME_USEC,
+        };
+
+        *ret = p;
+        return 0;
+}
+
+DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC(sd_radv_pref64_prefix, sd_radv_pref64_prefix, mfree);
+
+int sd_radv_pref64_prefix_set_prefix(
+                sd_radv_pref64_prefix *p,
+                const struct in6_addr *in6_addr,
+                unsigned char prefixlen,
+                uint32_t valid_lifetime) {
+
+        /* RFC 8781 - 4.1 rounding up lifetime to multiply of 8 */
+        uint32_t m1 = 0x00000000, m2 = 0x00000000;
+        uint8_t prefixlen_code;
+
+        assert_return(p, -EINVAL);
+        assert_return(in6_addr, -EINVAL);
+
+        if (prefixlen > 128)
+                return -EINVAL;
+
+        switch (prefixlen) {
+        case 96:
+                prefixlen_code = 0;
+                m1 = 0xffffffff;
+                m2 = 0xffffffff;
+                break;
+        case 64:
+                prefixlen_code = 1;
+                m1 = 0xffffffff;
+                break;
+        case 56:
+                prefixlen_code = 2;
+                m1 = 0xffffff00;
+                break;
+        case 48:
+                prefixlen_code = 3;
+                m1 = 0xffff0000;
+                break;
+        case 40:
+                prefixlen_code = 4;
+                m1 = 0xff000000;
+                break;
+        case 32:
+                prefixlen_code = 5;
+                break;
+        default:
+                log_radv(NULL, "Unusual PREF64 prefix length %u greater than 96", prefixlen);
```

Comment:
```suggestion
                log_radv(NULL, "Unsupported PREF64 prefix length %u", prefixlen);
```

Because we return `-EINVAL` here it is unsupported, not unusual. Also the "greater than" part does not apply to the default case.

Reply:
This was marked resolved, but remains unchanged.